HDDS-8132. [Design] Secure S3 keys management (#4372)
apache/ozone: hadoop-hdds/docs/content/design/secure-s3.md

---
title: Secure S3 keys management
summary: Improving security regarding s3 keys management
date: 2023-03-10
jira: HDDS-8132

> status: implemented

Reviewer: Just clarifying, a patch is available for remote S3 storage but, the implementation of HTTP endpoint for S3 secret retrieval is in progress correct?

Reviewer: Should we change the status to implementing then?

author: Maksim Myskov, Mikhail Pochatkin
---
<!--
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
   http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License. See accompanying LICENSE file.
-->

# Abstract

> 

Reviewer 2: _Add what is to be done:_
Extend Ozone S3 secret key management to support centralized remote S3 secret storage.  To add TTL to secrets.  To also implement S3 gateway endpoint to provide get s3 secret and secret rotataion.

There are problems related to the current S3 keys management:
* Storing keys as plain text in Ozone Manager rocksdb is insecure. An ozone administrator can easily impersonate any user by recovering his keys from rocksdb.
* The only way for a user to generate keys is to have SSH access to the Ozone cluster. Security policies can also prohibit this.
* Keys revocation process is manual which leads to security issues.

This document proposes solutions to the above issues.

# Link

> https://issues.apache.org/jira/secure/attachment/13056235/Secure%20S3%20keys%20management.pdf

Reviewer 2: In the design doc,
 
`Section 1. Store Secrets in remote storage`
last sentence, does loading all keys in ozone manager imply all keys are loaded to the proposed om cache for s3 secrets?  On om start, the om initiates a call to the (external) secret store interface to load all keys into cache?

`Section 2. Endpoint for key generation (renewal)`
Does the design propose to extend the s3g and add an endpoint or to create a separate s3 secret server endpoint with its own config (ozone.s3.secret...). ?

`Section 3. Add key expiration`
For key generation and renewal sharing the same gateway, defining and providing the REST API would be helpful.
What about key renewal?  Is there a provision for this?  Is it key generation with TTL, on s3g error retrying the s3 secret get to renew?

Reviewer 2: > providing the REST API

Apis, GET / POST parameters and responses / payload - Can we define these for the get secret and secret renewal an include them in the doc?

title: Secure S3 keys management
summary: Improving security regarding s3 keys management
date: 2023-03-10
jira: HDDS-8132
status: implementing
author: Maksim Myskov, Mikhail Pochatkin
---
<!--
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
   http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License. See accompanying LICENSE file.
-->

# Abstract

There are problems related to the current S3 keys management:
* Storing keys as plain text in Ozone Manager rocksdb is insecure. An ozone administrator can easily impersonate any user by recovering his keys from rocksdb.
* The only way for a user to generate keys is to have SSH access to the Ozone cluster. Security policies can also prohibit this.
* Keys revocation process is manual which leads to security issues.

> * Extend Ozone S3 secret key management to support centralized remote S3 secret storage

Reviewer 2: Thanks @myskov.  Can reword to, "We intend to extend Ozone S3 ...
                                                         Implement S3 gatewary endpoint ...
                                                         _And_ to add TTL to secrets.

jira: HDDS-8132
status: implementing
author: Maksim Myskov, Mikhail Pochatkin
---
<!--
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
   http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License. See accompanying LICENSE file.
-->

# Abstract

There are problems related to the current S3 keys management:
* Storing keys as plain text in Ozone Manager rocksdb is insecure. An ozone administrator can easily impersonate any user by recovering his keys from rocksdb.
* The only way for a user to generate keys is to have SSH access to the Ozone cluster. Security policies can also prohibit this.
* Keys revocation process is manual which leads to security issues.

We intend to extend Ozone S3 secret key management:
* Support centralized remote S3 secret storage.

> * Implement S3 gateway endpoint to provide get s3 secret and secret renewal.

Reviewer 3: S3 gateway endpoint we are adding even for revoke secret as well in HDDS-8050. We can add here along with get s3 secret, revoke secret.

Reviewer 4: ```suggestion
* Implement S3 gateway endpoint for getting, renewing and revoking secrets.
```

@ashishkumar50 @myskov please take a look at the suggested wording.  If it's OK, we can merge the PR after making this change.

Reviewer 3: @adoroszlai, Yes It is fine.
